keyv/Ceph.cpp

```diff
@@ -176,118 +78,272 @@ class Ceph : public Plugin
+    }
 
-        asyncRead.op.reset(librados::Rados::aio_create_completion());
-        uint64_t size = sizeHint;
-        if (size == 0)
+    pos = uri.findQuery("keyring");
+    if (pos != uri.queryEnd())
+    {
+        ret = _cluster.conf_set("keyring", pos->second.c_str());
+        if (ret < 0)
         {
-            time_t time;
-            const int stat = _context.stat(key, &size, &time);
-            if (stat < 0 || size == 0)
-            {
-                std::cerr << "Stat " << key << " failed: " << ::strerror(-stat)
-                          << std::endl;
-                return false;
-            }
+            _throw("Cannot read ceph keyring '" + pos->second + "'", ret);
         }
+    }
+
+    ret = _cluster.connect();
+    if (ret < 0)
+        _throw("Could not connect rados cluster", ret);
+
+    ret = _cluster.ioctx_create(poolName.c_str(), _context);
+    if (ret < 0)
+        _throw("Could not create io context", ret);
+}
 
-        const int read =
-            _context.aio_read(key, asyncRead.op.get(), &asyncRead.bl, size, 0);
-        if (read >= 0)
+inline Ceph::~Ceph()
+{
+    _context.close();
+    _cluster.shutdown();
+}
+
+inline bool Ceph::handles(const servus::URI& uri)
+{
+    return uri.getScheme() == "ceph";
+}
+
+inline std::string Ceph::getDescription()
+{
+    return "ceph://user@cluster?config=path&keyring=path";
+}
+
+inline size_t Ceph::setQueueDepth(const size_t depth)
+{
+    _maxPendingOps = depth;
+    // LBCHECK( _flush( _maxPendingOps ));
+    return _maxPendingOps;
+}
+
+inline bool Ceph::insert(const std::string& key, const void* data,
+                         const size_t size)
+{
+    librados::bufferlist bl;
+    bl.append((const char*)data, size);
+
+    if (_maxPendingOps == 0) // sync write
+    {
+        const int write = _context.write_full(key, bl);
+        if (write >= 0)
             return true;
 
-        std::cerr << "Fetch failed: " << ::strerror(-read) << std::endl;
+        std::cerr << "Write failed: " << ::strerror(-write) << std::endl;
         return false;
     }
 
-    bool _flush(const size_t maxPending)
+    librados::AioCompletion* op = librados::Rados::aio_create_completion();
+    const int write = _context.aio_write_full(key, op, bl);
+    if (write < 0)
     {
-        if (maxPending == 0)
+        std::cerr << "Write failed: " << ::strerror(-write) << std::endl;
+        delete op;
+        return false;
+    }
+    _writes.push_back(op);
+    return _flush(_maxPendingOps);
+}
+
+inline std::string Ceph::operator[](const std::string& key) const
+{
+    const librados::bufferlist bl = _get(key);
+    std::string value;
+    if (bl.length() > 0)
+        bl.copy(0, bl.length(), value);
+    return value;
+}
+
+inline void Ceph::takeValues(const lunchbox::Strings& keys,
+                             const ValueFunc& func) const
+{
+    size_t fetchedIndex = 0;
+    for (const auto& key : keys)
```

Comment:
Does rados have a multiget() like memcache? The fetch-and-get semantic here leaks from the 'old' SKV api and is less efficient since it needs more API calls..